source/fab/artefacts.py

```diff
@@ -4,35 +4,131 @@
 # which you should have received as part of this distribution
 ##############################################################################
 """
-This module contains :term:`Artefacts Getter` classes which return :term:`Artefact Collections <Artefact Collection>`
-from the :term:`Artefact Store`.
+This module contains :term:`Artefacts Getter` classes which return
+:term:`Artefact Collections <Artefact Collection>` from the
+:term:`Artefact Store`.
 
-These classes are used by the `run` method of :class:`~fab.steps.Step` classes to retrieve the artefacts
-which need to be processed. Most steps have sensible defaults and can be configured with user-defined getters.
+These classes are used by the `run` method of :class:`~fab.steps.Step`
+classes to retrieve the artefacts which need to be processed. Most steps
+have sensible defaults and can be configured with user-defined getters.
 
 """
+
 from abc import ABC, abstractmethod
+from collections import defaultdict
+from enum import auto, Enum
 from pathlib import Path
-from typing import Iterable, Union, Dict, List
+from typing import Dict, Iterable, List, Optional, Union
 
-from fab.constants import BUILD_TREES, CURRENT_PREBUILDS
 from fab.dep_tree import filter_source_tree, AnalysedDependent
 from fab.util import suffix_filter
 
 
+class ArtefactSet(Enum):
+    '''A simple enum with the artefact types used internally in Fab.
+    '''
+    ALL_SOURCE = auto()
+    PREPROCESSED_FORTRAN = auto()
+    PREPROCESSED_C = auto()
+    FORTRAN_BUILD_FILES = auto()
+    C_BUILD_FILES = auto()
+    X90_BUILD_FILES = auto()
+    CURRENT_PREBUILDS = auto()
+    PRAGMAD_C = auto()
+    BUILD_TREES = auto()
+    OBJECT_FILES = auto()
+    OBJECT_ARCHIVES = auto()
+    EXECUTABLES = auto()
+
+
 class ArtefactStore(dict):
-    '''This object stores artefacts (which can be of any type). Each artefact
+    '''This object stores set of artefacts (which can be of any type). Each artefact
     is indexed by a string.
     '''
+
     def __init__(self):
+        '''The constructor calls reset, which will mean all the internal
+        artefact categories are created.'''
         super().__init__()
         self.reset()
 
     def reset(self):
         '''Clears the artefact store (but does not delete any files).
         '''
         self.clear()
-        self[CURRENT_PREBUILDS] = set()
+        for artefact in ArtefactSet:
+            if artefact in [ArtefactSet.OBJECT_FILES,
+                            ArtefactSet.OBJECT_ARCHIVES]:
+                # ObjectFiles store a default dictionary (i.e. a non-existing
+                # key will automatically add an empty `set`)
+                self[artefact] = defaultdict(set)
+            else:
+                self[artefact] = set()
```

Comment:
Oh, I agree, I totally hated that :) My idea (for a different PR): we store two sets:
1. We map OBJECT_FILES (similar for OBJECT_ARCHIVES) to the set of root names (a list of strings)
2. We map each root name (string) to the set of object files required.

That would allow us to remove the special cases, but it will require to use a mixture of enums/strings as keys.

Of course, the trivial solution: each of these two sets just store a single dictionary, but that feels ugly as well. At least then the knowledge about what is stored there is restricted to the producer/consumer only (which already has this knowledge), it's not in ArtefactStore anymore.